blueflood-http/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/HttpMetricDataQueryServer.java

```diff
@@ -77,36 +76,52 @@ public void startServer() {
-        serverChannel =  (ServerChannel) server.bind(new InetSocketAddress(httpQueryHost, httpQueryPort));
+        ServerBootstrap server = new ServerBootstrap();
+        server.group(acceptorGroup, workerGroup)
+                .channel(NioServerSocketChannel.class)
+                .childHandler(new ChannelInitializer<SocketChannel>() {
+                    @Override
+                    public void initChannel(SocketChannel channel) throws Exception {
+                        setupPipeline(channel, finalRouter);
+                    }
+                });
+        serverChannel = server.bind(new InetSocketAddress(httpQueryHost, httpQueryPort)).sync().channel();
 
         //register the tracker MBean for JMX/jolokia
         log.info("Registering tracker service");
         Tracker.getInstance().register();
     }
 
-    private class MetricsHttpServerPipelineFactory implements ChannelPipelineFactory {
-        private RouteMatcher router;
-
-        public MetricsHttpServerPipelineFactory(RouteMatcher router) {
-            this.router = router;
-        }
-
-        @Override
-        public ChannelPipeline getPipeline() throws Exception {
-            final ChannelPipeline pipeline = pipeline();
-
-            pipeline.addLast("decoder", new HttpRequestDecoder());
-            pipeline.addLast("encoder", new HttpResponseEncoder());
-            pipeline.addLast("handler", new QueryStringDecoderAndRouter(router));
-
-            return pipeline;
-        }
+    private void setupPipeline(SocketChannel channel, RouteMatcher router) {
+        final ChannelPipeline pipeline = channel.pipeline();
+
+        pipeline.addLast("encoder", new HttpResponseEncoder());
+        pipeline.addLast("decoder", new HttpRequestDecoder() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable thr) throws Exception {
+                try {
+                    if (ctx.channel().isWritable()) {
+                        log.debug("request decoder error " + thr.getCause().toString() + " on channel " + ctx.channel().toString());
+                        ctx.channel().write(
+                                new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.BAD_REQUEST))
+                                            .addListener(ChannelFutureListener.CLOSE);
+                    } else {
+                        log.debug("channel " + ctx.channel().toString() + " is no longer writeable, not sending 400 response back to client");
+                    }
+                } catch (Exception ex) {
+                    // If we are getting exception trying to write,
+                    // don't propagate to caller. It may cause this
+                    // method to be called again and will produce
+                    // stack overflow. So just log it here.
+                    log.debug("Can't write to channel " + ctx.channel().toString(), ex);
```

Comment:
The main reason for switching from ```HttpRequest``` to ```FullHttpRequest``` is the ```HttpRequest``` interface has changed from Netty 3.x to Netty 4.x. 

In Netty 3.x, ```HttpRequest``` and ```HttpResponse``` are the only interfaces to use when dealing with Http requests and responses. From those interfaces, you can get the headers, the body, all necessary operations of Http, in those 2 interfaces.

In Netty 4.x, they have changed the interfaces to be smaller and more granular. ```HttpRequest``` only has methods to deal with headers -- that's all. You can't access the request body. To do that, you'll have to use ```FullHttpRequest``` interface.